### PR TITLE
[Android] Add initial extensible invoke command support

### PR DESCRIPTION
**Pull request by yunhanw-google**
ExtensibleInvoke can be used to send multiple invoke command at one time. In this PR.
-- Add extensibleInvoke and corresponding callback in Android Java/JNI
-- Add example test where we add two commands and send them out, and validate their response respectively.

### CODE REVIEW

**Review by joonhaengHeo — APPROVED** [review]

File: examples/java-matter-controller/java/src/com/matter/controller/commands/pairing/PairOnNetworkLongImBatchInvokeCommand.kt [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        if (status != null) {
+          setFailure("invoke failure with incorrect status")
+        }
+      }
+      responseCount++
+    }
+
+    override fun onNoResponse(noInvokeResponseData: NoInvokeResponseData) {
+      logger.log(Level.INFO, "Batch Invoke receive onNoResponse on $noInvokeResponseData")
+      responseCount--
```
Comment (tehampson) [inline examples/java-matter-controller/java/src/com/matter/controller/commands/pairing/PairOnNetworkLongImBatchInvokeCommand.kt:98]: I don't think you want to decrement here.

File: examples/java-matter-controller/java/src/com/matter/controller/commands/pairing/PairOnNetworkLongImBatchInvokeCommand.kt [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    val tlvWriter2 = TlvWriter()
+    tlvWriter2.startStructure(AnonymousTag)
+    tlvWriter2.put(ContextSpecificTag(0), number)
+    tlvWriter2.put(ContextSpecificTag(1), number)
+    tlvWriter2.endStructure()
+
+    val element2: InvokeElement =
+      InvokeElement.newInstance(
+        /* endpointId= */ 1,
+        CLUSTER_ID_TEST,
```
Comment (tehampson) [inline examples/java-matter-controller/java/src/com/matter/controller/commands/pairing/PairOnNetworkLongImBatchInvokeCommand.kt:146]: Is this test cluster there for all different servers?

I am trying to understand why we are deciding to send this second command here. Is this a temporary things, or is this expected to be more permanent?

File: src/controller/java/AndroidInteractionClient.cpp [hunk 3]
```diff
@@ -442,6 +442,195 @@ CHIP_ERROR PutPreencodedInvokeRequest(app::CommandSender & commandSender, app::C
...
+        isEndpointIdValid = (env->CallBooleanMethod(invokeElement, isEndpointIdValidMethod) == JNI_TRUE);
+        isGroupIdValid    = (env->CallBooleanMethod(invokeElement, isGroupIdValidMethod) == JNI_TRUE);
+
+        if (isEndpointIdValid)
+        {
+            endpointIdObj = env->CallLongMethod(invokeElement, getEndpointIdMethod, static_cast<jlong>(kInvalidEndpointId));
+            VerifyOrExit(!env->ExceptionCheck(), err = CHIP_JNI_ERROR_EXCEPTION_THROWN);
+        }
+
+        if (isGroupIdValid)
```
Comment (tehampson) [inline src/controller/java/AndroidInteractionClient.cpp:534]: This will always be false, right? Why is this here?

File: src/controller/java/AndroidCallbacks.h [hunk 4]
```diff
@@ -138,6 +139,24 @@ struct InvokeCallback : public app::CommandSender::Callback
...
+
+    void OnResponse(app::CommandSender * commandSender, const app::CommandSender::ResponseData & aResponseData) override;
+    void OnNoResponse(app::CommandSender * commandSender, const app::CommandSender::NoResponseData & aNoResponseData) override;
+    void OnError(const app::CommandSender * apCommandSender, const app::CommandSender::ErrorData & aErrorData) override;
+    void OnDone(app::CommandSender * apCommandSender) override;
+
+    CHIP_ERROR GetIndexForCommandRef(uint16_t aCommandRef, size_t & aIndex);
+    CHIP_ERROR AddCommandRefToIndexLookup(uint16_t aCommandRef, size_t aIndex);
+
+    std::unordered_map<uint16_t, size_t> commandRefToIndex;
```
Comment (tehampson) [inline src/controller/java/AndroidCallbacks.h:155]: If you don't want to, you don't have to use an unordered_map here. I would argue that is adds complexity without any benefit. This is something I am meaning to clean up in chip-repl.

In your case, because you are providing the `commandRef` to be the index already. You don't need this at all. If you want to be very safe, you can store the length of the array and validate that the CommandRef you get back in the OnResponse is less than the max. But you can rely on that `CommandRef` being the index for you.

File: src/controller/java/AndroidInteractionClient.cpp [hunk 5]
```diff
@@ -442,6 +442,195 @@ CHIP_ERROR PutPreencodedInvokeRequest(app::CommandSender & commandSender, app::C
     return writer->CopyContainer(TLV::ContextTag(app::CommandDataIB::Tag::kFields), reader);
 }
 
+CHIP_ERROR PutPreencodedInvokeRequest(app::CommandSender & commandSender, app::CommandPathParams & path, const ByteSpan & data,
+                                      app::CommandSender::PrepareCommandParameters & prepareCommandParams)
+{
+    // PrepareCommand does nott create the struct container with kFields and copycontainer below sets the
```
Comment (tehampson) [inline src/controller/java/AndroidInteractionClient.cpp:448]: nit, 
```suggestion
    // PrepareCommand does not create the struct container with kFields and copycontainer below sets the
```

File: examples/java-matter-controller/java/src/com/matter/controller/commands/pairing/PairOnNetworkLongImBatchInvokeCommand.kt (hunk 2 above)
Comment (yunhanw-google) [reply examples/java-matter-controller/java/src/com/matter/controller/commands/pairing/PairOnNetworkLongImBatchInvokeCommand.kt:146]: This is the test, we pair firstly, then send two command in batch, thanks

File: src/controller/java/AndroidInteractionClient.cpp [hunk 6]
```diff
@@ -442,6 +442,195 @@ CHIP_ERROR PutPreencodedInvokeRequest(app::CommandSender & commandSender, app::C
...
+    ChipLogDetail(Controller, "IM batchInvoke() called");
+
+    DeviceProxy * device = reinterpret_cast<DeviceProxy *>(devicePtr);
+    VerifyOrExit(device != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrExit(device->GetSecureSession().HasValue(), err = CHIP_ERROR_MISSING_SECURE_SESSION);
+
+    VerifyOrExit(invokeElementList != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
+    SuccessOrExit(err = JniReferences::GetInstance().GetListSize(invokeElementList, listSize));
+
+    if ((listSize > 1) && (device->GetSecureSession().Value()->IsGroupSession()))
```
Comment (tehampson) [inline src/controller/java/AndroidInteractionClient.cpp:482]: I don't think anything with group commands should ever call batch commands. They are two very different things

File: examples/java-matter-controller/java/src/com/matter/controller/commands/pairing/PairOnNetworkLongImBatchInvokeCommand.kt (hunk 2 above)
Comment (yunhanw-google) [reply examples/java-matter-controller/java/src/com/matter/controller/commands/pairing/PairOnNetworkLongImBatchInvokeCommand.kt:146]: Not permanent, this is temporary, we need the dedicated IM folder, and dedicated test suite that use IM for testing, that effort is still in radar , would happen in two or three months

File: src/controller/java/AndroidCallbacks.h (hunk 4 above)
Comment (yunhanw-google) [reply src/controller/java/AndroidCallbacks.h:155]: Agree, would remove it, thanks

File: src/controller/java/AndroidInteractionClient.cpp [hunk 7]
```diff
@@ -442,6 +442,195 @@ CHIP_ERROR PutPreencodedInvokeRequest(app::CommandSender & commandSender, app::C
...
+        app::CommandSender::FinishCommandParameters finishCommandParams(convertedTimedRequestTimeoutMs != 0
+                                                                            ? Optional<uint16_t>(convertedTimedRequestTimeoutMs)
+                                                                            : Optional<uint16_t>::Missing());
+
+        finishCommandParams.commandRef = prepareCommandParams.commandRef;
+        SuccessOrExit(err = commandSender->FinishCommand(finishCommandParams));
+
+        // CommandSender provides us with the CommandReference for this associated command. In order to match responses
+        // we have to add CommandRef to index lookup.
+        SuccessOrExit(err = callback->AddCommandRefToIndexLookup(finishCommandParams.commandRef.Value(), i));
```
Comment (tehampson) [inline src/controller/java/AndroidInteractionClient.cpp:598]: If you remove the `unordered_map` then you no longer actually need to track here

**Review by chrisdecenzo — APPROVED** [review]

File: src/controller/java/AndroidInteractionClient.cpp (hunk 6 above)
Comment (yunhanw-google) [reply src/controller/java/AndroidInteractionClient.cpp:482]: it would take care of all, updated with extenedableInvoke, thanks

File: src/controller/java/AndroidInteractionClient.cpp (hunk 7 above)
Comment (yunhanw-google) [reply src/controller/java/AndroidInteractionClient.cpp:598]: removed, thanks

File: src/controller/java/AndroidInteractionClient.cpp [hunk 8]
```diff
@@ -442,6 +442,195 @@ CHIP_ERROR PutPreencodedInvokeRequest(app::CommandSender & commandSender, app::C
...
+        }
+
+        app::CommandSender::FinishCommandParameters finishCommandParams(convertedTimedRequestTimeoutMs != 0
+                                                                            ? Optional<uint16_t>(convertedTimedRequestTimeoutMs)
+                                                                            : Optional<uint16_t>::Missing());
+
+        finishCommandParams.commandRef = prepareCommandParams.commandRef;
+        SuccessOrExit(err = commandSender->FinishCommand(finishCommandParams));
+
+        // CommandSender provides us with the CommandReference for this associated command. In order to match responses
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/controller/java/AndroidInteractionClient.cpp:596]: removed, thanks

File: src/controller/java/AndroidCallbacks.h (hunk 4 above)
Comment (yunhanw-google) [reply src/controller/java/AndroidCallbacks.h]: removed, thanks

File: src/controller/java/AndroidInteractionClient.cpp (hunk 5 above)
Comment (yunhanw-google) [reply src/controller/java/AndroidInteractionClient.cpp]: done, thanks

File: examples/java-matter-controller/java/src/com/matter/controller/commands/pairing/PairOnNetworkLongImBatchInvokeCommand.kt (hunk 1 above)
Comment (yunhanw-google) [reply examples/java-matter-controller/java/src/com/matter/controller/commands/pairing/PairOnNetworkLongImBatchInvokeCommand.kt]: thanks for catching. yes, removed, thanks

File: examples/java-matter-controller/java/src/com/matter/controller/commands/pairing/PairOnNetworkLongImBatchInvokeCommand.kt (hunk 2 above)
Comment (yunhanw-google) [reply examples/java-matter-controller/java/src/com/matter/controller/commands/pairing/PairOnNetworkLongImBatchInvokeCommand.kt]: it would be handled in https://github.com/project-chip/connectedhomeip/issues/25649, thanks

File: src/controller/java/AndroidInteractionClient.cpp (hunk 3 above)
Comment (yunhanw-google) [reply src/controller/java/AndroidInteractionClient.cpp:534]: it is for batch, and group...Update with extensible. thanks

**Review by tehampson — APPROVED** [review]

File: src/controller/java/AndroidCallbacks.h [hunk 9]
```diff
@@ -138,6 +139,20 @@ struct InvokeCallback : public app::CommandSender::Callback
     JniGlobalReference mWrapperCallbackRef;
 };
 
+struct ExtenedableInvokeCallback : public app::CommandSender::ExtendableCallback
```
Comment (andy31415) [inline src/controller/java/AndroidCallbacks.h:142]: seems like an extra `e` compared with the other other one:

`ExtendableCallback`
`ExtenedableInvokeCallback `

File: src/controller/java/MatterInteractionClient-JNI.cpp [hunk 10]
```diff
@@ -65,3 +65,15 @@ JNI_METHOD(void, invoke)
         ChipLogError(Controller, "JNI IM Invoke Error: %" CHIP_ERROR_FORMAT, err.Format());
     }
 }
+
+JNI_METHOD(void, extenedableInvoke)
```
Comment (andy31415) [inline src/controller/java/MatterInteractionClient-JNI.cpp:69]: extra E here as well. I think we should grep for `extenedable` and replace with `extendable` everywhere

File: src/controller/java/src/chip/devicecontroller/ExtenedableInvokeCallback.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+ *   limitations under the License.
+ *
+ */
+package chip.devicecontroller;
+
+import chip.devicecontroller.model.InvokeResponseData;
+import chip.devicecontroller.model.NoInvokeResponseData;
+
+/** An interface for receiving invoke response. */
+public interface ExtenedableInvokeCallback {
```
Comment (andy31415) [inline src/controller/java/src/chip/devicecontroller/ExtenedableInvokeCallback.java:24]: move file and rename `s/Extenedable/Extendable/`

File: src/controller/java/MatterInteractionClient-JNI.cpp (hunk 10 above)
Comment (yunhanw-google) [reply src/controller/java/MatterInteractionClient-JNI.cpp:69]: updated, thanks
